File: TypeScriptSource/Nolix/Core/Container/ListIterator.ts
//own import
import { List } from "./List";

//class
export class ListIterator<E> implements Iterator<E> {

    //attribute
    private listNode: ListNode<E>;

    //constructor
    constructor(listNode: ListNode<E>) {
        this.listNode = listNode;
    }

    //method
    next(): IteratorResult<E> {
    
        if (this.hasNext()) {
            let element = this.listNode.getRefElement();
            this.listNode = this.listNode.getRefNextNode();
            return { done: false, value: element };
        }

        return { done: true, value: null };
    }

    //method
    private hasNext(): boolean {
        return (this.listNode != null);
    }
}

File: Source/ch/nolix/common/independentcontainer/ListIterator.java
//package declaration
package ch.nolix.common.independentcontainer;

//Java import
import java.util.Iterator;

//own import
import ch.nolix.common.invalidargumentexception.ArgumentDoesNotHaveAttributeException;

//class
public final class ListIterator<E> implements Iterator<E> {

	//attribute
	private ListNode<E> nextNode;
	
	//constructor
	public ListIterator(final ListNode<E> nextNode) {
		this.nextNode = nextNode;
	}
	
	//method
	@Override
	public boolean hasNext() {
		return (nextNode != null);
	}

	//method
	@Override
	public E next() {
		
		if (!hasNext()) {
			throw new ArgumentDoesNotHaveAttributeException(this, "next element");
		}
		
		final E element = nextNode.getRefElement();
		
		nextNode = nextNode.getRefNextNodeOrNull();
		
		return element;
	}
}
